src/transformers/models/gemma2/modular_gemma2.py

```diff
@@ -480,27 +451,27 @@ def forward(
         all_hidden_states = () if output_hidden_states else None
         all_self_attns = () if output_attentions else None
 
-        for decoder_layer in self.layers[: self.config.num_hidden_layers]:
+        for i in range(self.config.num_hidden_layers):
             if output_hidden_states:
                 all_hidden_states += (hidden_states,)
 
             if self.gradient_checkpointing and self.training:
                 layer_outputs = self._gradient_checkpointing_func(
-                    partial(decoder_layer.__call__, **flash_attn_kwargs),
+                    partial(self.layers[i].__call__, **flash_attn_kwargs),
                     hidden_states,
                     position_embeddings,
-                    causal_mask,
+                    causal_masks[self.layer_attention_patterns[i]],
                     position_ids,
                     past_key_values,
                     output_attentions,
                     use_cache,
                     cache_position,
                 )
             else:
-                layer_outputs = decoder_layer(
+                layer_outputs = self.layers[i](
                     hidden_states,
                     position_embeddings=position_embeddings,
-                    attention_mask=causal_mask,
+                    attention_mask=causal_masks[self.layer_attention_patterns[i]],
```

Comment:
Let's put `self.layer_attention_patterns[i]` to access maybe layer.layer_type? WDYT